peps/pep-0749.rst

```diff
@@ -390,6 +390,138 @@ More specifically:
+class dictionary can render the descriptors invisible. Two approaches for fixing
+this issue are:
+
+* Ensure that the entry is *always* present in the class dictionary, even if it
+  is empty or has not yet been evaluated. This means we do not have to rely on
+  the descriptors defined on :py:class:`type` to fill in the field, and
+  therefore the metaclass's attributes will not interfere.
+* Ensure that the entry is *never* present in the class dictionary, or at least
+  never added by logic in the language core. This means that the descriptors
+  on :py:class:`type` will always be used, without interference from the metaclass.
+
+The first approach is straightforward to implement for ``__annotate__``: it is
+already always set for classes with annotations, and we can set it explicitly
+to ``None`` for classes without annotations.
+
+``__annotations__`` causes more problems, because it must be lazily evaluated.
+Therefore, we cannot simply always put an annotations dictionary in the class
+dictionary. The alternative approach would be to never set ``__dict__["__annotations__"]``
+and use some other storage to store the cached annotations. This behavior
+change would have to apply even to classes defined under
+``from __future__ import annotations``, because otherwise there could be buggy
+behavior if a class is defined without ``from __future__ import annotations``
+but its metaclass does have the future enabled. As :pep:`649` previously noted,
+removing ``__annotations__`` from class dictionaries also has backwards compatibility
+implications: ``cls.__dict__.get("__annotations__")`` is a common idiom to
+retrieve annotations.
+
+Alex Waygood has suggested an approach that avoids these problems. On class
+creation, ``cls.__dict__["__annotations__"]`` is set to a special descriptor.
+On ``__get__``, the descriptor evaluates the annotations by calling ``__annotate__``
+and replaces itself with the result. The descriptor also behaves like a mapping,
+so that code that uses ``cls.__dict__["__annotations__"]`` will still usually
+work: treating the object as a mapping will evaluate the annotations and behave
+as if the descriptor itself was the annotations dictionary. (Code that assumes
+that ``cls.__dict__["__annotations__"]`` is specifically an instance of ``dict``
+may break, however.)
+
+While this approach introduces significant complexity, it fixes all known
+problems with metaclasses and it preserves backwards compatibility better
+than alternative ideas.
+
+Specification
+-------------
+
+Accessing ``.__annotations__`` on a class object always returns that class's
+annotations dictionary, which may be empty. Accessing ``.__annotate__`` on a
+class object returns ``None`` if the class has no annotations, and an annotation
+function that returns the class's annotations if it does have annotations.
+
+Classes always contain ``__annotations__`` and ``__annotate__`` keys in their
```

Comment:
Does "classes" here mean all `type` instances? Or specifically heap types created with a `class` statement? What about the equivalent Python and C APIs?

I suspect only types created via a `class` statement getting these descriptors implicitly will make the most sense, with static types and types created via the dynamic Python and C APIs getting `__annotate__ = None` and `__annotations__ = {}` if nothing else is specified in the supplied class dicts.

Reply:
All classes that have >=1 class with annotations in their MRO must have `"__annotations__"` and `"__annotate__"` keys in their class dictionaries under this proposal. Theoretically I think it would be fine to omit these keys from the class dictionary if you could reliably determine that no classes in the MRO have any annotations. I don't know if that's worth the complexity, though.

Reply:
> All classes that have >=1 class with annotations in their MRO

Or in the MRO of a metaclass.

> Does "classes" here mean all type instances? Or specifically heap types created with a class statement? What about the equivalent Python and C APIs?

I think you're right and this should apply to classes created through the `class` statement, and classes created in ways that may inherit from Python classes should simply set the fields to None and {}.

Reply:
> > All classes that have >=1 class with annotations in their MRO
> 
> Or in the MRO of a metaclass.

right, yeah